File: Python3/Medium/Accepted/open-the-lock_3.py
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        
        deadends = set(deadends)
        if '0000' in deadends:
            return -1
        queue = collections.deque([('0000', 0)])
        visited = set()
        while queue:
            node, step = queue.popleft()
            if node == target:
                return step
            for i in range(4):
                for d in [-1, 1]:
                    new = node[:i] + str((int(node[i]) + d) % 10) + node[i+1:]
                    if new not in deadends and new not in visited:
                        queue.append((new, step + 1))
                        visited.add(new)
        return -1

File: C++/Medium/Accepted/open-the-lock_3.cpp
class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        
        unordered_set<string> dead(deadends.begin(), deadends.end());
        if (dead.count("0000")) return -1;
        unordered_set<string> visited;
        queue<string> q;
        q.push("0000");
        visited.insert("0000");
        int step = 0;
        while (!q.empty()) {
            int sz = q.size();
            for (int i = 0; i < sz; i++) {
                string cur = q.front(); q.pop();
                if (cur == target) return step;
                for (int j = 0; j < 4; j++) {
                    string up = cur, down = cur;
                    up[j] = up[j] == '9' ? '0' : up[j] + 1;
                    down[j] = down[j] == '0' ? '9' : down[j] - 1;
                    if (!visited.count(up) && !dead.count(up)) {
                        q.push(up);
                        visited.insert(up);
                    }
                    if (!visited.count(down) && !dead.count(down)) {
                        q.push(down);
                        visited.insert(down);
                    }
                }
            }
            step++;
        }
        return -1;
    }
};